File: app/src/main/java/org/blitzortung/android/data/provider/standard/JsonRpcDataProvider.java
package org.blitzortung.android.data.provider.standard;

import android.util.Log;

import org.blitzortung.android.app.Main;
import org.blitzortung.android.data.Parameters;
import org.blitzortung.android.data.beans.RasterParameters;
import org.blitzortung.android.data.beans.Station;
import org.blitzortung.android.data.beans.StrikeAbstract;
import org.blitzortung.android.data.provider.DataBuilder;
import org.blitzortung.android.data.provider.DataProvider;
import org.blitzortung.android.data.provider.DataProviderType;
import org.blitzortung.android.data.provider.result.ResultEvent;
import org.blitzortung.android.jsonrpc.JsonRpcClient;
import org.blitzortung.android.util.TimeFormat;
import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

import java.text.SimpleDateFormat;
import java.util.ArrayList;
import java.util.List;
import java.util.TimeZone;

public class JsonRpcDataProvider extends DataProvider {

    private static final SimpleDateFormat DATE_TIME_FORMATTER = new SimpleDateFormat("yyyyMMdd'T'HH:mm:ss");
    static private final String[] SERVERS = new String[]{"http://bo-service.tryb.de:7080/", "http://bo-service.tryb.de/"};
    static private int CURRENT_SERVER = 0;

    static {
        TimeZone tz = TimeZone.getTimeZone("UTC");
        DATE_TIME_FORMATTER.setTimeZone(tz);
    }

    private final DataBuilder dataBuilder;
    private JsonRpcClient client;
    private int nextId = 0;
    private boolean incrementalResult;

    public JsonRpcDataProvider() {
        dataBuilder = new DataBuilder();
    }

    private static String getServer() {
        return SERVERS[CURRENT_SERVER];
    }

    @Override
    public void getStrikes(Parameters parameters, ResultEvent.ResultEventBuilder result) {
        final int intervalDuration = parameters.getIntervalDuration();
        final int intervalOffset = parameters.getIntervalOffset();
        if (intervalOffset < 0) {
            nextId = 0;
        }
        result.incrementalData(nextId != 0);

        final int newStrikeCount;
        try {
            JSONObject response = client.call("get_strikes", intervalDuration, intervalOffset < 0 ? intervalOffset : nextId);

            newStrikeCount = addStrikes(response, result);
            addStrikesHistogram(response, result);
        } catch (Exception e) {
            skipServer();
            throw new RuntimeException(e);
        }

        Log.v(Main.LOG_TAG,
                String.format("JsonRpcDataProvider: read %d bytes (%d new strikes)", client.getLastNumberOfTransferredBytes(), newStrikeCount));
    }

    public boolean returnsIncrementalData() {
        return incrementalResult;
    }

    @Override
    public void getStrikesGrid(Parameters parameters, ResultEvent.ResultEventBuilder result) {
        nextId = 0;
        incrementalResult = false;

        final int intervalDuration = parameters.getIntervalDuration();
        final int intervalOffset = parameters.getIntervalOffset();
        final int rasterBaselength = parameters.getRasterBaselength();
        final int countThreshold = parameters.getCountThreshold();
        final int region = parameters.getRegion();

        final int elementCount;
        try {
            JSONObject response = client.call("get_strikes_grid", intervalDuration, rasterBaselength, intervalOffset, region, countThreshold);

            String info = String.format("%.0f km", rasterBaselength / 1000f);
            elementCount = addRasterData(response, result, info);
            addStrikesHistogram(response, result);
        } catch (Exception e) {
            skipServer();
            throw new RuntimeException(e);
        }

        Log.v(Main.LOG_TAG,
                String.format("JsonRpcDataProvider: read %d bytes (%d raster positions, region %d)", client.getLastNumberOfTransferredBytes(), elementCount, region));
    }

    @Override
    public List<Station> getStations(int region) {
        List<Station> stations = new ArrayList<>();

        try {
            JSONObject response = client.call("get_stations");
            JSONArray stations_array = (JSONArray) response.get("stations");

            for (int i = 0; i < stations_array.length(); i++) {
                stations.add(dataBuilder.createStation(stations_array.getJSONArray(i)));
            }
        } catch (Exception e) {
            skipServer();
            throw new RuntimeException(e);
        }
        return stations;
    }

    @Override
    public DataProviderType getType() {
        return DataProviderType.RPC;
    }

    @Override
    public void setUp() {
        String agentSuffix = pInfo != null ? "-" + Integer.toString(pInfo.versionCode) : "";
        client = new JsonRpcClient(getServer(), agentSuffix);
        client.setConnectionTimeout(40000);
        client.setSocketTimeout(40000);
    }

    @Override
    public void shutDown() {
        client.shutdown();
        client = null;
    }

    @Override
    public void reset() {
        nextId = 0;
    }

    @Override
    public boolean isCapableOfHistoricalData() {
        return true;
    }

    private int addStrikes(JSONObject response, ResultEvent.ResultEventBuilder result) throws JSONException {
        final List<StrikeAbstract> strikes = new ArrayList<>();
        long referenceTimestamp = getReferenceTimestamp(response);
        JSONArray strikes_array = (JSONArray) response.get("s");
        for (int i = 0; i < strikes_array.length(); i++) {
            strikes.add(dataBuilder.createDefaultStrike(referenceTimestamp, strikes_array.getJSONArray(i)));
        }
        if (response.has("next")) {
            nextId = (Integer) response.get("next");
        }
        result.strikes(strikes);
        return strikes.size();
    }

    private int addRasterData(JSONObject response, ResultEvent.ResultEventBuilder result, String info) throws JSONException {
        final RasterParameters rasterParameters = dataBuilder.createRasterParameters(response, info);
        long referenceTimestamp = getReferenceTimestamp(response);

        JSONArray strikes_array = (JSONArray) response.get("r");
        final List<StrikeAbstract> strikes = new ArrayList<>();
        for (int i = 0; i < strikes_array.length(); i++) {
            strikes.add(dataBuilder.createRasterElement(rasterParameters, referenceTimestamp, strikes_array.getJSONArray(i)));
        }

        result.rasterParameters(rasterParameters);
        result.strikes(strikes);

        return strikes.size();
    }

    private long getReferenceTimestamp(JSONObject response) throws JSONException {
        return TimeFormat.parseTime(response.getString("t"));
    }

    private void addStrikesHistogram(JSONObject response, ResultEvent.ResultEventBuilder result) throws JSONException {
        if (response.has("h")) {
            JSONArray histogram_array = (JSONArray) response.get("h");

            final int[] histogram = new int[histogram_array.length()];

            for (int i = 0; i < histogram_array.length(); i++) {
                histogram[i] = histogram_array.getInt(i);
            }
            result.histogram(histogram);
        }
    }

    private void skipServer() {
        CURRENT_SERVER = (CURRENT_SERVER + 1) % SERVERS.length;
    }
}

File: app/src/main/java/org/blitzortung/android/data/provider/standard/JsonRpcDataProvider.kt
/*

   Copyright 2015-2022 Andreas Würl

   Licensed under the Apache License, Version 2.0 (the "License");
   you may not use this file except in compliance with the License.
   You may obtain a copy of the License at

       http://www.apache.org/licenses/LICENSE-2.0

   Unless required by applicable law or agreed to in writing, software
   distributed under the License is distributed on an "AS IS" BASIS,
   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
   See the License for the specific language governing permissions and
   limitations under the License.

*/

package org.blitzortung.android.data.provider.standard

import android.content.SharedPreferences
import android.util.Log
import org.blitzortung.android.app.Main
import org.blitzortung.android.app.view.OnSharedPreferenceChangeListener
import org.blitzortung.android.app.view.PreferenceKey
import org.blitzortung.android.app.view.get
import org.blitzortung.android.data.Flags
import org.blitzortung.android.data.History
import org.blitzortung.android.data.Parameters
import org.blitzortung.android.data.beans.Station
import org.blitzortung.android.data.beans.Strike
import org.blitzortung.android.data.provider.DataProviderType
import org.blitzortung.android.data.provider.data.DataProvider
import org.blitzortung.android.data.provider.data.DataProvider.DataRetriever
import org.blitzortung.android.data.provider.data.initializeResult
import org.blitzortung.android.data.provider.result.ResultEvent
import org.blitzortung.android.jsonrpc.JsonRpcClient
import org.blitzortung.android.jsonrpc.JsonRpcResponse
import org.blitzortung.android.util.TimeFormat
import org.json.JSONArray
import org.json.JSONException
import org.json.JSONObject
import java.net.URI
import java.net.URL
import java.text.SimpleDateFormat
import java.util.*
import javax.inject.Inject
import javax.inject.Singleton

@Singleton
class JsonRpcDataProvider @Inject constructor(
    preferences: SharedPreferences,
    private val client: JsonRpcClient
) : OnSharedPreferenceChangeListener, DataProvider {

    private lateinit var serviceUrl: URL
    private val dataBuilder: DataBuilder = DataBuilder()
    private var nextId = 0

    init {
        preferences.registerOnSharedPreferenceChangeListener(this)
        onSharedPreferenceChanged(preferences, PreferenceKey.SERVICE_URL)

        Log.v(Main.LOG_TAG, "JsonRpcDataProvider($serviceUrl)")
    }

    override val type: DataProviderType = DataProviderType.RPC

    override fun reset() {
        nextId = 0
    }

    @Throws(JSONException::class)
    private fun addStrikes(response: JsonRpcResponse, result: ResultEvent): ResultEvent {
        val isIncremental = nextId != 0
        val strikes = ArrayList<Strike>()
        val referenceTimestamp = getReferenceTimestamp(response.data)
        val strikesArray = response.data.get("s") as JSONArray
        for (i in 0 until strikesArray.length()) {
            strikes.add(dataBuilder.createDefaultStrike(referenceTimestamp, strikesArray.getJSONArray(i)))
        }
        if (response.data.has("next")) {
            nextId = response.data.get("next") as Int
        }
        return result.copy(strikes = strikes, updated = if (isIncremental) strikes.size else -1)
    }

    private fun addGridData(response: JsonRpcResponse, result: ResultEvent, size: Int): ResultEvent {
        val gridParameters = dataBuilder.createGridParameters(response.data, size)
        val referenceTimestamp = getReferenceTimestamp(response.data)

        val strikesArray = response.data.get("r") as JSONArray
        val strikes = ArrayList<Strike>()
        for (i in 0 until strikesArray.length()) {
            strikes.add(
                dataBuilder.createGridElement(
                    gridParameters,
                    referenceTimestamp,
                    strikesArray.getJSONArray(i)
                )
            )
        }

        return result.copy(strikes = strikes, gridParameters = gridParameters, referenceTime = referenceTimestamp)
    }

    @Throws(JSONException::class)
    private fun getReferenceTimestamp(response: JSONObject): Long {
        return TimeFormat.parseTime(response.getString("t"))
    }

    @Throws(JSONException::class)
    private fun addStrikesHistogram(response: JSONObject, result: ResultEvent): ResultEvent {
        var resultVar = result

        if (response.has("h")) {
            val histogramArray = response.get("h") as JSONArray

            val histogram = IntArray(histogramArray.length())

            for (i in 0 until histogramArray.length()) {
                histogram[i] = histogramArray.getInt(i)
            }
            resultVar = resultVar.copy(histogram = histogram)
        }

        return resultVar
    }

    override fun <T> retrieveData(retrieve: DataRetriever.() -> T): T = Retriever(client).retrieve()


    private inner class Retriever(val client: JsonRpcClient) : DataRetriever {
        override fun getStations(region: Int): List<Station> {
            val stations = ArrayList<Station>()

            try {
                val response = client.call(serviceUrl, "get_stations")
                val stationsArray = response.data.get("stations") as JSONArray

                for (i in 0 until stationsArray.length()) {
                    stations.add(dataBuilder.createStation(stationsArray.getJSONArray(i)))
                }
            } catch (e: Exception) {
                throw RuntimeException(e)
            }

            return stations
        }

        override fun getStrikesGrid(parameters: Parameters, history: History?, flags: Flags): ResultEvent {
            var result = initializeResult(parameters, history, flags)

            nextId = 0

            try {
                val response = JsonRpcData(client, serviceUrl).requestData(parameters)
                result = addGridData(response, result, parameters.gridSize)
                result = addStrikesHistogram(response.data, result)
            } catch (e: Exception) {
                throw RuntimeException(e)
            }

            Log.v(
                Main.LOG_TAG,
                "JsonRpcDataProvider: read %d bytes (%d grid entries, region %d)".format(
                    client.lastNumberOfTransferredBytes,
                    result.strikes?.size,
                    parameters.region
                )
            )
            return result
        }

        override fun getStrikes(parameters: Parameters, history: History?, flags: Flags): ResultEvent {
            var result = initializeResult(parameters, history, flags)

            val intervalDuration = parameters.intervalDuration
            val intervalOffset = parameters.intervalOffset
            if (intervalOffset < 0) {
                nextId = 0
            }

            try {
                val response = client.call(
                    serviceUrl,
                    "get_strikes",
                    intervalDuration,
                    if (intervalOffset < 0) intervalOffset else nextId
                )

                result = addStrikes(response, result)
                result = addStrikesHistogram(response.data, result)
            } catch (e: Exception) {
                throw RuntimeException(e)
            }

            Log.v(
                Main.LOG_TAG,
                "JsonRpcDataProvider: read %d bytes (%d new strikes)".format(
                    client.lastNumberOfTransferredBytes,
                    result.strikes?.size
                )
            )
            return result
        }
    }

    override fun onSharedPreferenceChanged(sharedPreferences: SharedPreferences, key: PreferenceKey) {
        when (key) {
            PreferenceKey.SERVICE_URL -> {
                val serviceUrl: String =
                    sharedPreferences.get(PreferenceKey.SERVICE_URL, DEFAULT_SERVICE_URL.toString())
                this.serviceUrl =
                    toCheckedUrl(if (serviceUrl.isNotBlank()) serviceUrl.trim() else DEFAULT_SERVICE_URL.toString())
            }
            else -> {}
        }
    }

    private fun toCheckedUrl(serviceUrl: String): URL {
        return try {
            URL(serviceUrl)
        } catch (e: Exception) {
            Log.e(Main.LOG_TAG, "JsonRpcDataProvider.tocheckedUrl($serviceUrl) invalid")
            DEFAULT_SERVICE_URL
        }
    }

    companion object {
        private val DATE_TIME_FORMATTER = SimpleDateFormat("yyyyMMdd'T'HH:mm:ss", Locale.US)
        private val DEFAULT_SERVICE_URL = URI("http://bo-service.tryb.de/").toURL()

        init {
            val tz = TimeZone.getTimeZone("UTC")
            DATE_TIME_FORMATTER.timeZone = tz
        }
    }
}
